### PR TITLE
fix(python): dont raise when partial function is passed to map_elements

### LINKED ISSUE #13523: Series.map_elements raises an AttributeError with partial functions
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
from functools import partial

def plus(value, amount: int):
        return value + amount

data = {"a": [1, 2], "b": [3, 4]}
df = pl.DataFrame(data)
applied = df["a"].map_elements(partial(plus, amount=1))
print(applied)
```

### Log output

```shell
>       applied = df["a"].map_elements(partial(plus, amount=1))
  
.venv/lib/python3.12/site-packages/polars/series/series.py:5260: in map_elements
    warn_on_inefficient_map(function, columns=[self.name], map_target="series")
.venv/lib/python3.12/site-packages/polars/utils/udfs.py:890: in warn_on_inefficient_map
    module, suggestion = _is_raw_function(function)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

function = functools.partial(<function test_udf.<locals>.plus at 0x104a25ee0>, amount=1)

    def _is_raw_function(function: Callable[[Any], Any]) -> tuple[str, str]:
        """Identify translatable calls that aren't wrapped inside a lambda/function."""
        func_module = function.__class__.__module__
>       func_name = function.__name__
E       AttributeError: 'functools.partial' object has no attribute '__name__'. Did you mean: '__ne__'?

.venv/lib/python3.12/site-packages/polars/utils/udfs.py:840: AttributeError
```

### Issue description

partial functions do not have a __name__ attribute.
This worked until polars 0.20.2 but is broken as of 0.20.3-rc1

### Expected behavior

No exception and the partial function is applied to all elements 

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.3
Index type:           UInt32
Platform:             macOS-14.2.1-arm64-arm-64bit
Python:               3.12.0 (main, Oct  2 2023, 12:03:24) [Clang 15.0.0 (clang-1500.0.40.1)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
numpy:                <not installed>
openpyxl:             <not installed>
pandas:               <not installed>
pyarrow:              <not installed>
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #13523

this regressed in https://github.com/pola-rs/polars/pull/13221 , sorry about that - but hey, increased test coverage now

### CODE REVIEW

**Review by alexander-beedie — APPROVED** [review]

**Comment by alexander-beedie** [comment]
Looks good to me, thanks! :)

File: py-polars/tests/unit/operations/map/test_inefficient_map_warning.py [hunk 1]
```diff
@@ -406,3 +407,13 @@ def test_expr_exact_warning_message() -> None:
...
+
+
+def test_partial_functions_13523() -> None:
+    def plus(value, amount: int):  # type: ignore[no-untyped-def]
+        return value + amount
+
+    data = {"a": [1, 2], "b": [3, 4]}
+    df = pl.DataFrame(data)
+    # should not warn
+    _ = df["a"].map_elements(partial(plus, amount=1))
```
Comment (stinodego) [inline py-polars/tests/unit/operations/map/test_inefficient_map_warning.py:419]: Let's actually make sure this doesn't warn?

```suggestion
def test_partial_functions_13523(recwarn: Any) -> None:
    def plus(value, amount: int):  # type: ignore[no-untyped-def]
        return value + amount

    data = {"a": [1, 2], "b": [3, 4]}
    df = pl.DataFrame(data)
    _ = df["a"].map_elements(partial(plus, amount=1))
    assert len(recwarn) == 0
```

**Review by stinodego — COMMENTED** [review]
Thanks for the fix! Just one comment about the test.

File: py-polars/tests/unit/operations/map/test_inefficient_map_warning.py (hunk 1 above)
Comment (stinodego) [reply py-polars/tests/unit/operations/map/test_inefficient_map_warning.py:419]: I was too late 😄 but it's fine, don't worry about it.

**Comment by robvanmieghem** [comment]
Thank you  for the fast fix

**Pull request merged**